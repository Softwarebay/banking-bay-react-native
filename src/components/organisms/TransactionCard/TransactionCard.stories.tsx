import { color, number, text } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react-native';
import { CenterStory } from 'utils/storybook/CenterStory';

import { TransactionCard } from './TransactionCard';

storiesOf('Organisms/TransactionCard', module)
	.addDecorator((getStory) => <CenterStory>{getStory()}</CenterStory>)
	.add('Outgoing transaction', () => (
		<TransactionCard
			amount={number('amount', 140.25)}
			category={text('subtitle', 'Money Transfer')}
			iconColor={color('iconColor', '#FF8A71')}
			iconName="smartphone"
			id="story-transaction-outgoing"
			title={text('title', 'Jane Doe')}
			type="outgoing"
		/>
	))
	.add('Incoming transaction', () => (
		<TransactionCard
			amount={number('amount', 140.25)}
			category={text('subtitle', 'Money Transfer')}
			iconColor={color('iconColor', '#FF8A71')}
			iconName="smartphone"
			id="story-transaction-incoming"
			title={text('title', 'Jane Doe')}
			type="incoming"
		/>
	));
