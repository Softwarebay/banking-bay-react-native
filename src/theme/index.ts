interface Colors {
	screenBackground: string;
	mainDark: string;
	bodyText: string;
	pink: string;
	buttonLight: string;
	buttonDark: string;
	green: string;
	white: string;
	shadow: string;
	grey600: string;
	grey300: string;
	grey100: string;
}

export const colors: Colors = {
	screenBackground: '#FAFCFF',
	mainDark: '#1B1D4D',
	bodyText: '#4C4C60',
	pink: '#FF5887',
	buttonLight: '#97DAFF',
	buttonDark: '#16579D',
	green: '#3EB290',
	white: '#FFFFFF',
	shadow: '#161966',
	grey600: '#868698',
	grey300: '#CED6E1',
	grey100: '#F1F5FD',
};

interface Shadows {
	ios: {
		shadowColor: string;
		shadowOffset: {
			width: number;
			height: number;
		};
		shadowOpacity: number;
		shadowRadius: number;
	};
	android: {
		elevation: number;
		shadowColor: string;
	};
}

export const shadows: Shadows = {
	ios: {
		shadowColor: colors.shadow,
		shadowOffset: {
			width: 0,
			height: 6,
		},
		shadowOpacity: 0.08,
		shadowRadius: 16,
	},
	android: {
		elevation: 4,
		shadowColor: colors.shadow,
	},
};
