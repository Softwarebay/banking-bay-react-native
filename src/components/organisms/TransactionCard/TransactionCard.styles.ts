import { StyleSheet } from 'react-native';
import { colors } from 'theme';

export const styles = StyleSheet.create({
	amount: {},
	container: {
		paddingHorizontal: 10,
		paddingVertical: 19,
	},
	description: {
		flex: 1,
		marginHorizontal: 14,
	},
	iconContainer: {
		alignItems: 'center',
		backgroundColor: colors.transactionCardIcon,
		borderRadius: 100,
		height: 40,
		justifyContent: 'center',
		width: 40,
	},

	subtitle: {
		fontSize: 12,
		lineHeight: 19.2,
	},
});
