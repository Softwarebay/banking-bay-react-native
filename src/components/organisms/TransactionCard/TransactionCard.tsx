import { StyleProp, View, ViewStyle } from 'react-native';

import { Card, Heading, Icon, IconProps, Text } from 'components/atoms';

import { colors } from 'theme';
import { styles } from './TransactionCard.styles';

interface Props {
	id: string;
	iconName: IconProps['name'];
	iconColor: string;
	type: 'incoming' | 'outgoing';
	title: string;
	category: string;
	amount: number;
	style?: StyleProp<ViewStyle>;
}

export const TransactionCard = ({
	id,
	iconName,
	iconColor,
	title,
	category,
	type,
	amount,
	style,
}: Props) => {
	const amountText = `${type === 'incoming' ? '+' : '-'} ${amount}`;
	const amountColor = `${
		type === 'incoming' ? colors.success : colors.mainDark
	}`;

	return (
		<Card style={[styles.container, style]}>
			<View style={styles.iconContainer}>
				<Icon
					color={iconColor}
					name={iconName}
					testID={`${id}-transaction-icon-test-id`}
				/>
			</View>
			<View style={styles.description}>
				<Heading
					size="h6"
					testID={`${id}-transaction-title-test-id`}
					text={title}
				/>
				<Text
					style={styles.category}
					testID={`${id}-transaction-subtitle-test-id`}
					text={category}
				/>
			</View>
			<Heading
				color={amountColor}
				size="h6"
				testID={`${id}-transaction-amount-test-id`}
				text={amountText}
			/>
		</Card>
	);
};
