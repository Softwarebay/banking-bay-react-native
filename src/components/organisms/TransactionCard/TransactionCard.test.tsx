import { fireEvent, render, screen } from '@testing-library/react-native';
import testRenderer from 'react-test-renderer';

import { colors } from 'theme';
import { TransactionCard } from './TransactionCard';

import type { Props } from './TransactionCard';

describe('TransactionCard', () => {
	const mockOnPress = jest.fn<Props['onPress'], []>();
	const props: Omit<Props, 'type'> = {
		onPress: mockOnPress,
		testID: 'transaction-card-test-id',
		iconName: 'smartphone',
		iconColor: 'blue',
		amount: 10.0,
		category: 'Mobile Transfer',
		id: 'mobile',
		title: 'Transaction Title',
		style: {
			backgroundColor: 'blue',
		},
	};

	it('should match snapshot - incoming transfer', () => {
		const tree = testRenderer
			.create(<TransactionCard {...props} type="incoming" />)
			.toJSON();

		expect(tree).toMatchSnapshot();
	});

	it('should match snapshot - outgoing transfer', () => {
		const tree = testRenderer
			.create(<TransactionCard {...props} type="outgoing" />)
			.toJSON();

		expect(tree).toMatchSnapshot();
	});

	it('should render icon, title, category subtitle, amount of money transferred', () => {
		render(<TransactionCard {...props} type="outgoing" />);

		const icon = screen.getByTestId('smartphone-icon-test-id');
		const title = screen.getByTestId('mobile-transaction-title-test-id');
		const subtitle = screen.getByTestId('mobile-transaction-subtitle-test-id');
		const amount = screen.getByTestId('mobile-transaction-amount-test-id');

		expect(icon).not.toBeNull();
		expect(title).toHaveTextContent('Transaction Title');
		expect(subtitle).toHaveTextContent('Mobile Transfer');
		expect(amount).toHaveTextContent(/10.00/);
	});

	it("should display incoming money transfer as green text with '+' sign", () => {
		render(<TransactionCard {...props} type="incoming" />);

		const amount = screen.getByTestId('mobile-transaction-amount-test-id');

		expect(amount).toHaveStyle({
			color: colors.green,
		});
		expect(amount).toHaveTextContent('+ 10.00');
	});

	it("should display outgoing money transfer as default color text with '-' sign", () => {
		render(<TransactionCard {...props} type="outgoing" />);

		const amount = screen.getByTestId('mobile-transaction-amount-test-id');

		expect(amount).toHaveStyle({
			color: colors.mainDark,
		});
		expect(amount).toHaveTextContent('- 10.00');
	});

	it('should apply additional styles', () => {
		render(<TransactionCard {...props} type="outgoing" />);

		const [card] = screen.getByTestId('transaction-card-test-id').children;

		expect(card).toHaveStyle({
			backgroundColor: 'blue',
		});
	});

	it('should handle press events', () => {
		render(<TransactionCard {...props} type="outgoing" />);

		const pressableCard = screen.getByTestId('transaction-card-test-id');

		fireEvent.press(pressableCard);

		expect(mockOnPress).toHaveBeenCalledTimes(1);
	});
});
