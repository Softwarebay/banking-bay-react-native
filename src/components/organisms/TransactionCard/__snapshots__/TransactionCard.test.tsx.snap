// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransactionCard should match snapshot - incoming transfer 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  testID="transaction-card-test-id"
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
        },
        Object {
          "flexDirection": "row",
        },
        Array [
          Object {
            "backgroundColor": "#FFFFFF",
            "borderRadius": 10,
            "shadowColor": "#161966",
            "shadowOffset": Object {
              "height": 6,
              "width": 0,
            },
            "shadowOpacity": 0.08,
            "shadowRadius": 16,
          },
          Array [
            Object {
              "paddingHorizontal": 15,
              "paddingVertical": 15,
            },
            Object {
              "backgroundColor": "blue",
            },
            false,
          ],
        ],
      ]
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#F1F5FD",
          "borderRadius": 100,
          "height": 50,
          "justifyContent": "center",
          "width": 50,
        }
      }
    >
      <SvgMock
        color="blue"
        height={25}
        testID="smartphone-icon-test-id"
        width={25}
      />
    </View>
    <View
      style={
        Object {
          "flex": 1,
          "marginHorizontal": 15,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#1B1D4D",
            },
            Object {
              "fontFamily": "MulishSemiBold",
              "fontSize": 16,
              "lineHeight": 25.6,
            },
            undefined,
          ]
        }
        testID="mobile-transaction-title-test-id"
      >
        Transaction Title
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#4C4C60",
              "fontFamily": "MulishRegular",
              "fontSize": 17,
              "lineHeight": 27.2,
            },
            Object {
              "fontSize": 14,
              "lineHeight": 22.4,
            },
          ]
        }
        testID="mobile-transaction-subtitle-test-id"
      >
        Mobile Transfer
      </Text>
    </View>
    <Text
      style={
        Array [
          Object {
            "color": "#3EB290",
          },
          Object {
            "fontFamily": "MulishSemiBold",
            "fontSize": 16,
            "lineHeight": 25.6,
          },
          undefined,
        ]
      }
      testID="mobile-transaction-amount-test-id"
    >
      + 10.00
    </Text>
  </View>
</View>
`;

exports[`TransactionCard should match snapshot - outgoing transfer 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  testID="transaction-card-test-id"
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
        },
        Object {
          "flexDirection": "row",
        },
        Array [
          Object {
            "backgroundColor": "#FFFFFF",
            "borderRadius": 10,
            "shadowColor": "#161966",
            "shadowOffset": Object {
              "height": 6,
              "width": 0,
            },
            "shadowOpacity": 0.08,
            "shadowRadius": 16,
          },
          Array [
            Object {
              "paddingHorizontal": 15,
              "paddingVertical": 15,
            },
            Object {
              "backgroundColor": "blue",
            },
            false,
          ],
        ],
      ]
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#F1F5FD",
          "borderRadius": 100,
          "height": 50,
          "justifyContent": "center",
          "width": 50,
        }
      }
    >
      <SvgMock
        color="blue"
        height={25}
        testID="smartphone-icon-test-id"
        width={25}
      />
    </View>
    <View
      style={
        Object {
          "flex": 1,
          "marginHorizontal": 15,
        }
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#1B1D4D",
            },
            Object {
              "fontFamily": "MulishSemiBold",
              "fontSize": 16,
              "lineHeight": 25.6,
            },
            undefined,
          ]
        }
        testID="mobile-transaction-title-test-id"
      >
        Transaction Title
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#4C4C60",
              "fontFamily": "MulishRegular",
              "fontSize": 17,
              "lineHeight": 27.2,
            },
            Object {
              "fontSize": 14,
              "lineHeight": 22.4,
            },
          ]
        }
        testID="mobile-transaction-subtitle-test-id"
      >
        Mobile Transfer
      </Text>
    </View>
    <Text
      style={
        Array [
          Object {
            "color": "#1B1D4D",
          },
          Object {
            "fontFamily": "MulishSemiBold",
            "fontSize": 16,
            "lineHeight": 25.6,
          },
          undefined,
        ]
      }
      testID="mobile-transaction-amount-test-id"
    >
      - 10.00
    </Text>
  </View>
</View>
`;
